lib/js/jslib.js

```diff
@@ -800,21 +800,76 @@ const LINKS = new (function() {
       return null;
     }
 
-    if (obj.func) {
-      const f = (!TYPES.isObject(obj.environment)) ?
-        eval(obj.func) :
-        partialApply(eval(obj.func), eval(obj.environment));
-      f.location = obj.location; // This may be set to 'server' by the server serializer.
-      f.func = obj.func;
-      return f;
-    } else if (obj.key) {
-      obj.key = _lookupMobileKey(state, obj.key)
-      return obj;
-    } else {
-      for (let key in obj) {
-        obj[key] = resolveServerValue(state, obj[key]);
-      }
-      return obj;
+    switch (obj._tag) {
+      // Primatives
+      case "Bool":
+      case "Int":
+      case "Float":
+      case "Database":
+      case "Table":
+      case "XML":
+      case "String":
+        return obj._val;        
+      
+      case "Char":
+        delete obj._tag; 
+        return obj;
+
+      case "FunctionPtr":
+      case "ClientFunction":
+        const f = (!TYPES.isObject(obj.environment)) ?
+          eval(obj.func) :
+          partialApply(eval(obj.func), eval(obj.environment));
+        f.location = obj.location; // This may be set to 'server' by the server serializer.
+        f.func = obj.func;
+        return f;
+
+      case "Record":
+        var newObj = {};
+        for (let k in obj) {
+          newObj[k] = resolveServerValue(state, obj[k]);
+        }
+        return newObj;
+      case "Variant":
+        var value = {};
+        var newObj = { '_label': obj._label, '_value': value};
+
+        for (let k in obj._value) {
+          value[k] = resolveServerValue(state, obj[k]);
+        }
+        return newObj;
+        
+      case "List":
+        if (obj._head === null) {
+          return {} // check
+        }
+        return {
+          '_head': resolveServerValue(state, obj._head),
+          '_tail': resolveServerValue(state, obj._tail)
+        };
+      
+      case "ClientAccessPoint":
+      case "ServerAccessPoint":
+      case "ClientPid":
+      case "ServerPid":
+      case "SessionChannel":
+      case "ServerSpawnLoc":
+      case "ClientupawnLoc":
+        // TODO: should they return a new object?
+        delete obj._tag;
+        return obj;
+
+      default:
+        delete obj._tag;
+        if (obj.key) {
+          obj.key = _lookupMobileKey(state, obj.key)
+          return obj;
+        } else {
+          for (let k in obj) {
```

Comment:
This branch should be deleted.

Reply:
`_lookupMobileKey` is no longer used in the function. Is that fine?